chapters/part3/hyperparameter-optimization-with-optuna.md

```diff
@@ -0,0 +1,362 @@
+local lr = std.parseJson(std.extVar('lr'));
+local max_filter_size = std.parseInt(std.extVar('max_filter_size'));
+local num_filters = std.parseInt(std.extVar('num_filters'));
+local output_dim = std.parseInt(std.extVar('output_dim'));
+local ngram_filter_sizes = std.range(2, max_filter_size);
+```
+
+That's it. You can view a final configuration by clicking `details` below.
+
+<details>
+<br>
+
+`imdb_optuna.jsonnet`
+
+```json
+local batch_size = 64;
+local cuda_device = 0;
+local num_epochs = 15;
+local seed = 42;
+local train_data_path = 'https://s3-us-west-2.amazonaws.com/allennlp/datasets/imdb/train.jsonl';
+local validation_data_path = 'https://s3-us-west-2.amazonaws.com/allennlp/datasets/imdb/dev.jsonl';
+
+local embedding_dim = std.parseInt(std.extVar('embedding_dim'));
+local dropout = std.parseJson(std.extVar('dropout'));
+local lr = std.parseJson(std.extVar('lr'));
+local max_filter_size = std.parseInt(std.extVar('max_filter_size'));
+local num_filters = std.parseInt(std.extVar('num_filters'));
+local output_dim = std.parseInt(std.extVar('output_dim'));
+local ngram_filter_sizes = std.range(2, max_filter_size);
+
+{
+  numpy_seed: seed,
+  pytorch_seed: seed,
+  random_seed: seed,
+  dataset_reader: {
+    type: 'text_classification_json',
+    tokenizer: {
+      type: 'whitespace',
+    },
+    token_indexers: {
+      tokens: {
+        type: 'single_id',
+      },
+    },
+  },
+  datasets_for_vocab_creation: ['train'],
+  train_data_path: train_data_path,
+  validation_data_path: validation_data_path,
+  model: {
+    type: 'basic_classifier',
+    text_field_embedder: {
+      token_embedders: {
+        tokens: {
+          embedding_dim: embedding_dim,
+        },
+      },
+    },
+    seq2vec_encoder: {
+      type: 'cnn',
+      embedding_dim: embedding_dim,
+      ngram_filter_sizes: ngram_filter_sizes,
+      num_filters: num_filters,
+      output_dim: output_dim,
+    },
+    dropout: dropout,
+  },
+  data_loader: {
+    batch_size: batch_size,
+  },
+  trainer: {
+    cuda_device: cuda_device,
+    num_epochs: num_epochs,
+    optimizer: {
+      lr: lr,
+      type: 'sgd',
+    },
+    validation_metric: '+accuracy',
+  },
+}
+```
+
+</details>
+
+## II: Defining Search Space for Hyperparameters in Pythonic way
+
+Now that you have created the config, the next step is defining the hyperparameter search spaces.
+In Optuna, a search space is defined by creating an `objective function`.
+Each hyperparameter search space is declared with `suggest_int` or `suggest_float`.
+For categorical hyperparameters, you can use `suggest_categorical`.
+Please see <a href="https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial">our documentation</a> for more information.
+In this tutorial, I create an objective function like following (Note that the names of the hyperparameters should be the same as those defined in the configuration earlier):
```

Comment:
The names of the hyperparameters are not clearly explained here.
How about adding a short explanation about the argument of suggest API.

The following is just an example:
`These suggest functions require two kinds of arguments at least. The first one is the name of hyperparameter, and the second one is the range of the values. Note that the name of the ...`

Reply:
I adopted your suggestion in ae5cd52.

Reply:
Thank you!